[curl] Fixes pkgconfig configuration file

Matioupi opened the pull request:
- What does your PR fix?
This PR fixes the libcurl.pc pkgconfig configuration file for windows debug build. It also enables tool on non windows platforms (except uwp)

- Which triplets are supported/not supported? Have you updated the CI baseline?

test triplets : x64-windows / x64-windows-static / x86-windows / x86-windows-static / x64-linux / arm64-linux
(tested with/without tool feature enabled) and x64-uwp (core feature only)

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes, to the best of my knowledge...

Matioupi committed:
Fix pkgconfig for windows debug build. Re-allow tool feature on non windows platforms

Matioupi committed:
Fix tool feature

PhoebeHui commented on ports/curl/portfile.cmake:
```diff
@@ -109,5 +113,14 @@ else()
     )
 endif()
 
+# Fix the pkgconfig file for debug
+if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+    if(VCPKG_TARGET_IS_WINDOWS)
+        vcpkg_replace_string(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc "-lcurl" "-lcurl-d")
+        vcpkg_replace_string(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc "/lib/zlib.lib" "/debug/lib/zlibd.lib")	
+    endif()
+    file(COPY ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig)
+endif()
```
Could you use vcpkg_fixup_pkgconfig() instead?

Matioupi replied:
Hello, I tried, but without working examples of how to properly us it, I only achieved to get error messages such as :

for additionnal command :

vcpkg_fixup_pkgconfig(DEBUG_FILES libcurl.pc)

`-- Fixing pkgconfig file: C:/src/vcpkg/packages/curl_x64-windows/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_curl C:/src/vcpkg/packages/curl_x64-windows/lib/zlib.lib_RELEASE:CHECK_LIB_curl C:/src/vcpkg/packages/curl_x64-windows/lib/zlib.lib_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:232 (message):
  Library "curl C:/src/vcpkg/packages/curl_x64-windows/lib/zlib.lib" was not
  found! If it is a system library use the SYSTEM_LIBRARIES parameter for the
  vcpkg_fixup_pkgconfig call! Otherwise, correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:298 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:125 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)`

So in the end, I used a search/replace strategy (as in the zlib portfile by the way).

If you have some examples of how to use vcpkg_fixup_pkgconfig() properly (e.g. in this particular case) I'd change it in the PR and can also add the exemple in https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_fixup_pkgconfig.md which do not contain specific exemples.

PhoebeHui replied:
@Matioupi, it seems vcpkg_fixup_pkgconfig() doesn't handle this case, cc @Neumann-A, he might be interested in this.

I don't have a good suggestion now, I will approve the changes.

Neumann-A replied:
`vcpkg_fixup_pkgconfig()` only fixes the paths to make them work with relative paths and checks if the libraries mentioned within the pc file exists. If they do not exists they need to be corrected. As such a) first correct the library names in the *.pc file and b) call vcpkg_fixup_pkgconfig() after that to handle/correct the paths and make them relative. 
vcpkg_fixup_pkgconfig is not intended to automatically fix -l flags!

PhoebeHui replied:
@Neumann-A, I think the zlib also need to be replaced.

From
Libs.private:  -lwldap32 -lwinmm -lws2_32 -ladvapi32 -lcrypt32 E:/vcpkg/clean/vcpkg/installed/x64-windows/lib/zlib.lib

To
Libs.private:  -lwldap32 -lwinmm -lws2_32 -ladvapi32 -lcrypt32 -lzlib

Then it will check the zlib.pc file when use vcpkg_fixup_pckconfig.

Neumann-A commented on ports/curl/portfile.cmake:
```diff
@@ -109,5 +113,14 @@ else()
+    endif()
+    file(COPY ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig)
+endif()
+
```
add the call after you corrected the *.pc files
```suggestion
vcpkg_fixup_pkgconfig()
```

Matioupi replied:
Hello, I took into account both your request and the one from @PhoebeHui to change :/vcpkg/clean/vcpkg/installed/x64-windows/lib/zlib.lib style lib to -lzlib. Actually, I had to do it otherwise vcpkg_fixup_pkgconfig() call was failing.
I also added other libs for some feature enabled builds such as :
vcpkg.exe install curl[core,tool,http2,ssh,openssl,winssl,mbedtls,sspi]:x64-windows

Matioupi committed:
Modified debug/release prvate libs according to https://github.com/microsoft/vcpkg/pull/14350#discussion_r520444862
Added vcpkg_fixup_pkgconfig() according to https://github.com/microsoft/vcpkg/pull/14350#discussion_r520435718

Matioupi committed:
Merge remote-tracking branch 'upstream/master' into curl_fixes

Matioupi committed:
Delete portfile.cmake.orig

Commit:
Fix pkgconfig

Matioupi committed:
Revert "Fix pkgconfig"

This reverts commit cb4da8bcc5e92248c20065f34d5e5d1d88a0e5d4.

Matioupi commented:
With commit https://github.com/microsoft/vcpkg/commit/cb4da8bcc5e92248c20065f34d5e5d1d88a0e5d4
build with features such as : .\vcpkg\vcpkg.exe install curl[core,tool,http2,ssh,openssl,winssl,mbedtls,sspi]:x64-windows
fails (where not failing before the commit)

Matioupi committed:
Bump port version to 3

PhoebeHui commented:
@Matioupi, you need to add the following libs as system libraries likes the commit 'cb4da8b' do, I suggest to test this changes locally with defalut and features


See the CI test failures:

x86-windows/x64-windows/x64-windows-static/arm64-windows
```
-- Fixing pkgconfig file: D:/packages/curl_x86-windows/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_wldap32_RELEASE:CHECK_LIB_wldap32_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "wldap32" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:144 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)
```
x64-uwp:
```
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "crypt32" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:144 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)
```

x64-linux and x64-osx
```
-- Fixing pkgconfig file: /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_c /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libssl.a /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libcrypto.a /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libz.a_RELEASE:CHECK_LIB_c /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libssl.a /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libcrypto.a /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libz.a_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "c
  /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libssl.a
  /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libcrypto.a
  /mnt/vcpkg-ci/packages/curl_x64-linux/lib/pkgconfig/../../lib/libz.a" was
  not found! If it is a system library use the SYSTEM_LIBRARIES parameter for
  the vcpkg_fixup_pkgconfig call! Otherwise, correct the *.pc file
```

Matioupi committed:
add missing system libs for windows. made case type coherent for ws2_32

Matioupi commented:
@PhoebeHui : Wouldn't it be better to add those libs as Windows system libs "vcpkg wide" ? as proposed in : https://github.com/microsoft/vcpkg/pull/14350/commits/aee37bd36685ab0571aa5e0a7a22fdf2e4866d4e ?

Matioupi committed:
Fix build on x64-linux triplet

Neumann-A commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -192,5 +192,7 @@ if(VCPKG_TARGET_IS_WINDOWS)
     list(APPEND VCPKG_SYSTEM_LIBRARIES winmm)
     list(APPEND VCPKG_SYSTEM_LIBRARIES wsock32)
-    list(APPEND VCPKG_SYSTEM_LIBRARIES Ws2_32)
+    list(APPEND VCPKG_SYSTEM_LIBRARIES ws2_32)
```
NO! Add both cases if necessary. vcpkg_fixup_pkgconfig does not do a case insensitive compare!

Matioupi replied:
Ok, so far I've seen no side effet of having ws2_32 instead of Ws2_32 and this change was mostly cosmetic. I believe it's better to have only once the Ws2_32 with the capital W than twice. Will revert to Ws2_32

Matioupi committed:
Reverted ws2_32 to Ws2_32 as discudssed in https://github.com/microsoft/vcpkg/pull/14350#discussion_r525893479

Matioupi committed:
More fixes for Linux build with features

PhoebeHui commented:
@Matioupi, the feature 'c-ares' test fail with x64-windows, could you please take a look?

```
-- CHECK_LIB_curl E:/vcpkg/14350/vcpkg/packages/curl_x64-windows/lib/cares.lib_RELEASE:CHECK_LIB_curl E:/vcpkg/14350/vcpkg/packages/curl_x64-windows/lib/cares.lib_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "curl E:/vcpkg/14350/vcpkg/packages/curl_x64-windows/lib/cares.lib"
  was not found! If it is a system library use the SYSTEM_LIBRARIES parameter
  for the vcpkg_fixup_pkgconfig call! Otherwise, correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:166 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)
```

Error: Building package curl:x64-windows failed with: BUILD_FAILED

Matioupi committed:
Fix for feature c-ares

Matioupi committed:
Formatting

Matioupi commented:
Should be fixed now. I guess the regressions in C.I. are unrelated to those commits

PhoebeHui commented:
@Matioupi, please ignore it, currently it has been fixed, I have rerun the CI testing.

All features test passed with x64-windows.

strega-nil commented:
Alright, thanks @Matioupi :D
